politeiawww/email.go

```diff
@@ -0,0 +1,482 @@
+// proposal becoming censored to the proposal's author.
+func (b *backend) emailAuthorForCensoredProposal(
+	proposal *v1.ProposalRecord,
+	authorUser *database.User,
+	adminUser *database.User,
+) error {
+	if b.cfg.SMTP == nil {
+		return nil
+	}
+
+	l, err := url.Parse(b.cfg.WebServerAddress + "/proposals/" +
+		proposal.CensorshipRecord.Token)
+	if err != nil {
+		return err
+	}
+
+	if authorUser.ProposalEmailNotifications&
+		uint64(v1.NotificationEmailMyProposalStatusChange) == 0 {
+		return nil
+	}
+
+	tplData := proposalStatusChangeTemplateData{
+		Link:               l.String(),
+		Name:               proposal.Name,
+		StatusChangeReason: proposal.StatusChangeMessage,
+	}
+
+	subject := "Your Proposal Has Been Censored"
+	body, err := createBody(templateProposalCensoredForAuthor, &tplData)
+	if err != nil {
+		return err
+	}
+
+	return b.sendEmailTo(subject, body, authorUser.Email)
+}
+
+// emailUsersForVettedProposal sends an email notification for a new
+// proposal becoming vetted.
+func (b *backend) emailUsersForVettedProposal(
+	proposal *v1.ProposalRecord,
+	authorUser *database.User,
+	adminUser *database.User,
+) error {
+	if b.cfg.SMTP == nil {
+		return nil
+	}
+
+	// Create the template data.
+	l, err := url.Parse(b.cfg.WebServerAddress + "/proposals/" +
+		proposal.CensorshipRecord.Token)
+	if err != nil {
+		return err
+	}
+
+	tplData := proposalStatusChangeTemplateData{
+		Link:     l.String(),
+		Name:     proposal.Name,
+		Username: authorUser.Username,
+	}
+
+	// Send email to users.
+	subject := "New Proposal Published"
+	body, err := createBody(templateProposalVetted, &tplData)
+	if err != nil {
+		return err
+	}
+
+	return b.sendEmail(subject, body, func(msg *goemail.Message) error {
+		// Add user emails
+		return b.db.AllUsers(func(user *database.User) {
+			// Don't notify the user under certain conditions.
+			if user.NewUserPaywallTx == "" || user.Deactivated ||
+				user.ID == adminUser.ID || user.ID == authorUser.ID ||
+				(user.ProposalEmailNotifications&
+					uint64(v1.NotificationEmailRegularProposalVetted)) == 0 {
+				return
+			}
+
+			msg.AddBCC(user.Email)
+		})
+	})
+}
+
+// emailUsersForEditedProposal sends an email notification for a proposal
+// being edited.
+func (b *backend) emailUsersForEditedProposal(
+	proposal *v1.ProposalRecord,
+	authorUser *database.User,
+) error {
+	if b.cfg.SMTP == nil {
+		return nil
+	}
+
+	// Create the template data.
+	l, err := url.Parse(b.cfg.WebServerAddress + "/proposals/" +
+		proposal.CensorshipRecord.Token)
+	if err != nil {
+		return err
+	}
+
+	tplData := proposalEditedTemplateData{
+		Link:     l.String(),
+		Name:     proposal.Name,
+		Version:  proposal.Version,
+		Username: authorUser.Username,
+	}
+
+	// Send email to users.
+	subject := "Proposal Edited"
+	body, err := createBody(templateProposalEdited, &tplData)
+	if err != nil {
+		return err
+	}
+
+	return b.sendEmail(subject, body, func(msg *goemail.Message) error {
```

Comment:
This is really hard to read. At least document what is going on.

Reply:
I documented the sendEmail() function as well as improved the docs for each call.